zio-http/src/main/scala/zio/http/QueryParams.scala

```diff
@@ -26,41 +29,31 @@ import zio.http.internal.QueryParamEncoding
 /**
  * A collection of query parameters.
  */
-final case class QueryParams(map: Map[String, Chunk[String]]) {
-  self =>
+trait QueryParams {
+  self: QueryParams =>
+
+  /**
+   * All query parameters as a map
+   */
+  def map: Map[String, Chunk[String]]
 
   /**
    * Combines two collections of query parameters together. If there are
    * duplicate keys, the values from both sides are preserved, in order from
    * left-to-right.
    */
-  def ++(that: QueryParams): QueryParams =
-    QueryParams(that.map.foldLeft(map) { case (map, (k, v2)) =>
-      map.updated(
-        k,
-        map.get(k) match {
-          case Some(v1) => v1 ++ v2
-          case None     => v2
-        },
-      )
-    })
+  def ++(that: QueryParams): QueryParams
 
   /**
    * Adds the specified key/value pair to the query parameters.
    */
-  def add(key: String, value: String): QueryParams = addAll(key, Chunk(value))
+  def add(key: String, value: String): QueryParams =
+    addAll(key, Chunk(value))
 
   /**
    * Adds the specified key/value pairs to the query parameters.
    */
-  def addAll(key: String, value: Chunk[String]): QueryParams = {
-    val previousValue = map.get(key)
-    val newValue      = previousValue match {
-      case Some(prev) => prev ++ value
-      case None       => value
-    }
-    QueryParams(map.updated(key, newValue))
-  }
+  def addAll(key: String, value: Chunk[String]): QueryParams
```

Comment:
I would not leave this abstract. Provide a default implementation in terms of `++` and the `QueryParam.apply` constructor.

Reply:
Good call, done.